tools/wpt/browser.py

```diff
@@ -939,44 +944,214 @@ class Chrome(ChromeChromiumBase):
+            raise requests.RequestException(
+                "Chrome for Testing versions not found", e)
+        versions_json = resp.json()
+        versions_list = versions_json["versions"]
+        # Attempt to find a version match in the list of available downloads.
+        matching_versions = [v for v in versions_list if v["version"] == version]
+        if len(matching_versions) == 0:
+            raise DownloadNotFoundError(f"No Chrome for Testing download found for {version}")
+
+        download_info = matching_versions[0]["downloads"]
+        # Find the download url that matches the current platform.
+        browser_download_urls = [d for d in download_info["chrome"]
+                                 if d["platform"] == self._chrome_platform_string]
+        if len(browser_download_urls) == 0:
+            raise DownloadNotFoundError(
+                f"No Chrome for Testing download found for {self.platform} of version {version}")
+        browser_download_url = browser_download_urls[0]["url"]
+
+        # Get the corresponding ChromeDriver download URL for later use.
+        chromedriver_download_urls = [d for d in download_info["chromedriver"]
+                                      if d["platform"] == self._chrome_platform_string]
+        if len(chromedriver_download_urls) == 0:
+            # Some older versions of Chrome for Testing
+            # do not have a matching ChromeDriver download.
+            raise DownloadNotFoundError(
+                f"ChromeDriver download does not exist for version {version}")
+        chromedriver_url = chromedriver_download_urls[0]["url"]
+
+        return browser_download_url, chromedriver_url
+
+    def _get_download_urls_by_channel(self, channel):
+        """Find Chrome for Testing and ChromeDriver download URLs matching the given channel.
+
+        Returns: Both binary downloads for Chrome and ChromeDriver.
+        """
+        try:
+            resp = get(
+                "https://googlechromelabs.github.io"
+                "/chrome-for-testing/last-known-good-versions-with-downloads.json")
+        except requests.RequestException as e:
+            raise requests.RequestException(
+                "Chrome for Testing versions not found", e)
+        channels_json = resp.json()
+        download_info = channels_json["channels"][channel.capitalize()]["downloads"]["chrome"]
+
+        # Find the download URL that matches the current platform.
+        matching_download_urls = [d for d in download_info
+                                  if d["platform"] == self._chrome_platform_string]
+        if len(matching_download_urls) == 0:
+            raise DownloadNotFoundError("No matching download for platform "
+                                        f"{self.platform} of channel \"{channel}\".")
+
+        browser_download_url = matching_download_urls[0]["url"]
+
+        # Get the corresponding ChromeDriver download URL for later use.
+        chromedriver_download_info = (
```

Comment:
@mathiasbynens is the schema for returning in the last-known-good-versions-with-downloads JSON guaranteed to be backwards compatible?

I'm worried if we need to add some additional checks to prevent KeyError exceptions in the future if the schema breaks.

Reply:
It is true that this will not work to match ChromeDriver versions < 115 I believe. If we want to continue to allow users to supply their own version of Chrome, and attempt to find a download of ChromeDriver to match that older version, it's worthwhile to keep the old ChromeDriver download source as a fallback. I'll add a case for that.

Reply:
> @mathiasbynens is the schema for returning in the last-known-good-versions-with-downloads JSON guaranteed to be backwards compatible?

Yes